aws/resource_aws_datasync_location_smb.go

```diff
@@ -0,0 +1,247 @@
+	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
+)
+
+func resourceAwsDataSyncLocationSmb() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsDataSyncLocationSmbCreate,
+		Read:   resourceAwsDataSyncLocationSmbRead,
+		Update: resourceAwsDataSyncLocationSmbUpdate,
+		Delete: resourceAwsDataSyncLocationSmbDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"agent_arns": {
+				Type:     schema.TypeSet,
+				Required: true,
+				ForceNew: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+			"domain": {
+				Type:     schema.TypeString,
+				Computed: true,
+				Optional: true,
+			},
+			"mount_options": {
+				Type:     schema.TypeList,
+				Optional: true,
+				ForceNew: true,
+				MaxItems: 1,
+				// Ignore missing config block (stolen from aws_ecs_service)
+				DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
+					if old == "1" && new == "0" {
+						return true
+					}
+					return false
+				},
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"version": {
+							Type:         schema.TypeString,
+							Default:      "AUTOMATIC",
+							Optional:     true,
+							ForceNew:     true,
+							ValidateFunc: validation.NoZeroValues,
+						},
+					},
+				},
+			},
+			"password": {
+				Type:      schema.TypeString,
+				Required:  true,
+				Sensitive: true,
+			},
+			"server_hostname": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"subdirectory": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				/*// Ignore missing trailing slash
+				DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
+					if new == "/" {
+						return false
+					}
+					if strings.TrimSuffix(old, "/") == strings.TrimSuffix(new, "/") {
+						return true
+					}
+					return false
+				},
+				*/
+			},
+			"tags": {
+				Type:     schema.TypeMap,
+				Optional: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+			"uri": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"user": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+		},
+	}
+}
+
+func resourceAwsDataSyncLocationSmbCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).datasyncconn
+
+	input := &datasync.CreateLocationSmbInput{
+		AgentArns:      expandStringSet(d.Get("agent_arns").(*schema.Set)),
+		MountOptions:   expandDataSyncSmbMountOptions(d.Get("mount_options").([]interface{})),
+		Password:       aws.String(d.Get("password").(string)),
+		ServerHostname: aws.String(d.Get("server_hostname").(string)),
+		Subdirectory:   aws.String(d.Get("subdirectory").(string)),
+		Tags:           expandDataSyncTagListEntry(d.Get("tags").(map[string]interface{})),
```

Comment:
Can you please refactor the tagging code to use the new `keyvaluetags` package - https://github.com/terraform-providers/terraform-provider-aws/issues/10688?